### DISCUSSION TITLE
How to call native function from typescript

### DISCUSSION THREAD

**Discussion by xmeow** [microsoft/devicescript#650] [Q&A] [answered]
I have a native function in the firmware, and I want to call it from typescript. Here are two possible ways I managed to do it.

1. Add a new custom service and register the native function as a characteristic. This will allow you to call the native function from typescript using the characteristic write method. The drawback is that you have to write a lot of boilerplate code to implement the service and characteristic, and it is very slow to call the native function.

2. Modify the `bytecode.md` to add customized instructions. This will allow you to call the native function directly, like the `ledStripSend` function and make implementation of the native function in the firmware. But such way will break the compatibility and there will always an limitation of the number of bytecodes you can use. And each firmware will have their own customized bytecodes.

Or someother ways I don't know. Thank you for your help.

**Comment by pelikhan** [comment]
@mmoskal

**Comment by mmoskal** [comment, marked as answer]
This are pretty much the only two ways, and you correctly identified the trade-offs. For the 1st one you can implement a Jacdac action, not a property, but it's pretty much the same thing. It is the recommended way for use cases where speed is not critical.

Note that similar interpreters, like MicroPython, only have the second option.

One thing that we could do to improve the second option, is to allow arbitrary strings to refer to runtime methods and objects. That would mean you wouldn't have to update `bytecode.md` but could instead check for existence of a method/object in normal JavaScript ways and fall back to some other behavior. However, there are no immediate plans to do it and I'm not sure it would help your use case much...

**Comment by xmeow** [comment]
Thank you for taking the time to share your insights. While I understand there are no immediate plans for its implementation, it's an interesting perspective to consider. 

Once again, thank you for your guidance. Looking forward to further collaboration.

**Comment by pelikhan** [comment]
One idea would be to support an APU similar to what we do for SPI currently: buffer in, buffer out and let the c++ layer listeners handle the packets. This avoid having to design JacDac services for low level primitive and would not have JacDac packet limitation

**Comment by iBobik** [comment]
Hello,
I am trying to use display not supported by Devicescript itself, but supported by the [LovyanGFX](https://github.com/lovyan03/LovyanGFX) library (C++).

So I had idea to try to make an universal display driver for Devicescript, but I struggle with wiring Devicescript project with the C++ code.

It seems reasonable to go the JacDac way, but I found no examples how to write "native function as a characteristic“. I am also not sure how to configure Devicescript project to install C library and compile the native code, so it can be called from the JacDac service.

It there something like this implemented already? Thank you for help.
